thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/function/AbstractModularizedAnomalyFunction.java

```diff
@@ -34,19 +28,13 @@
  * with the given anomaly detection module; the actions can be anomaly detection, information update
  * of merged anomalies, etc.
  */
-public abstract class AbstractModularizedAnomalyFunction extends BaseAnomalyFunction implements AnomalyDetectionFunction {
+public abstract class AbstractModularizedAnomalyFunction extends BaseAnomalyFunction implements AnomalyDetectionFunction,
+    ModularizedAnomalyFunctionModelProvider {
   protected final Logger LOGGER = LoggerFactory.getLogger(this.getClass());
 
   protected AnomalyFunctionDTO spec;
   protected Properties properties;
 
-  protected DataModel dataModel = new NoopDataModel();
```

Comment:
you can still keep them here, just make them private. In the init method, call getDataModel(), getPredictionModel() etc and set these variables.